### ISSUE TITLE
Bazel 7: cannot build protobuf on Linux without --noincompatible_sandbox_hermetic_tmp

### ISSUE THREAD

**Issue by stapelberg** [bazelbuild/bazel#20621]
### Description of the bug:

I cannot build the protobuf repository on Linux anymore with Bazel 7:

```
/tmp/protobuf main % bazel build :protoc //conformance:conformance_test_runner               
INFO: Analyzed 2 targets (165 packages loaded, 2175 targets configured).
ERROR: /tmp/protobuf/src/google/protobuf/BUILD.bazel:194:11: Compiling src/google/protobuf/port.cc failed: (Exit 1): gcc failed: error executing CppCompile command (from target //src/google/protobuf:port_def) /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer '-std=c++14' -MD -MF ... (remaining 22 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
src/google/protobuf/port.cc:12:10: fatal error: google/protobuf/port_def.inc: No such file or directory
   12 | #include "google/protobuf/port_def.inc"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 1.193s, Critical Path: 0.30s
INFO: 239 processes: 233 internal, 6 linux-sandbox.
ERROR: Build did NOT complete successfully

```

This issue seems to only manifest when using the `linux-sandbox` strategy. When switching to `processwrapper-sandbox`, the build works. When adding `--noincompatible_sandbox_hermetic_tmp`, the build works.

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

1. `git clone https://github.com/protocolbuffers/protobuf`
2. `cd protobuf`
3. `bazel build :protoc //conformance:conformance_test_runner`

### Which operating system are you running Bazel on?

Linux

### What is the output of `bazel info release`?

release 7.0.0

**Comment by fmeum** [comment]
@bazel-io fork 7.0.1

**Comment by fmeum** [comment]
@lberki I tested this with `last_green` and it still happens if the workspace source directory lies under `/tmp`.

**Comment by lberki** [comment]
https://github.com/bazelbuild/bazel/pull/20603 fixes this for me, both for the case where only the output base is under `/tmp` and for the case where both the workspace and the output base are there.

@fmeum or @stapelberg mind trying it out? It'll be a while before that can get merged (I can't finish the work today, I can't find a reviewer tomorrow and I'm OOO from the day after) so it'd be nice to know if it's *another* issue or a known one. I'm getting close to cutting my losses and flipping `--incompatible_sandbox_hermetic_tmp` back and if this is a new issue, that'd probably convince me to do so.

**Comment by fmeum** [comment]
It's possible that I did something wrong, but tested this with a Bazel built from d1ab02e6142ae6039ed730031f5ea1ae6e334ec1 with protobuf checked out into `/tmp/protobuf` and `/tmp/git/protobuf`, both with default output bases and an output base of `/tmp/proto-out`, and am still seeing the failure.

**Comment by lberki** [comment]
Did you patch it on top of HEAD? (I haven't rebased it so it doesn't contain https://github.com/bazelbuild/bazel/commit/3748084889d85b132e0a8371cb2ab1eafdc311da yet)

**Comment by fmeum** [comment]
I tried this now and it still fails for me. Which commands are you running exactly?

**Comment by lberki** [comment]
```
git clone https://github.com/protocolbuffers/protobuf.git
cd protobuf
ghbazel --output_base=/tmp/ob build :protoc  # ghbazel is development version
```

You?

**Comment by fmeum** [comment]
When I run these commands after `cd ~/git`, they pass, when I run them after `cd /tmp/git`, they fail.

**Comment by lberki** [comment]
Gotcha, you're right -- I have `--disk_cache=` set by default. Disabling that and having the source checkout tickles the bug.

**Comment by lberki** [comment]
Update: the latest version of #20603 fixes the above (unfortunately, it's *still* not an easy change to land)

**Comment by lberki** [comment]
With fb6658c86164b81205a056a9a54975a62f2f957a submitted, I'll mark this as fixed. Let me know if it still doesn't work with Bazel after that change!

**Comment by sluongng** [comment]
Our team ran into this issue yesterday as well.

Applying `--noincompatible_sandbox_hermetic_tmp` workaround resolved the issue.

Looking forward to 7.0.1 🙏

**Comment by iancha1992** [comment]
Cherry-picked in https://github.com/bazelbuild/bazel/pull/20718. Just FYI @sluongng @lberki

**Comment by iancha1992** [comment]
@bazel-io fork 7.1.0

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.0.1 RC2](https://releases.bazel.build/7.0.1/rc2/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!
